### ISSUE TITLE
Tf-idf calculation

### ISSUE THREAD

**Issue by dgrtwo** [juliasilge/tidytext#11]
Tf-idf (term frequency-inverse document frequency) can be calculated for a one-term-per-document-per-row tidy dataset using a few group_by and mutate statements, but we should add this as one or more verbs since it's a common operation.

One way I could imagine it would be a function `d %>% tf_idf(term, document, n)` that adds three columns:
- `tf` for term frequency (n / total in doc)
- `idf` for inverse document frequency (log(n_distinct(document) / # of docs with term))
- `tf_idf` for the product of the above

There are variations that we'd have to explore and add options for

**Comment by dgrtwo** [comment]
Saving a draft here for later:

``` r
#' Calculate the term-frequency and inverse document frequency
#' of a tidy text dataset
#'
#' Calculate the term frequency and inverse document frequency of a
#' tidy text dataset, along with the product, tf-idf. Each of these
#' values are added as columns.
#'
#' @param tbl A tidy text dataset with one-row-per-term-per-document
#' @param term_col Column containing terms
#' @param document_col Column containing document IDs
#' @param n_col Column containing document-term counts
#'
#' @details \code{tf_idf} is given bare names, while \code{tf_idf_}
#' is given strings and is therefore suitable to programming with.
#'
#' If the dataset is grouped, the groups are ignored but are
#' retained.
#'
#' @examples
#'
#' library(dplyr)
#' library(janeaustenr)
#'
#' book_words <- austen_books() %>%
#'   unnest_tokens(word, text) %>%
#'   count(book, word, sort = TRUE)
#'
#' book_words
#'
#' # find the words most distinctive to each document
#' book_words %>%
#'   tf_idf(word, book, n) %>%
#'   arrange(desc(tf_idf))
#'
#' @export
tf_idf <- function(tbl, term_col, document_col, n_col) {
  tf_idf_(tbl,
          col_name(substitute(term_col)),
          col_name(substitute(document_col)),
          col_name(substitute(n_col)))
}


#' @rdname tf_idf
#' @export
tf_idf_ <- function(tbl, term_col, document_col, n_col) {
  g <- groups(tbl)

  total_documents <- n_distinct(tbl[[document_col]])

  dots <- list(.document_total = substitute(sum(x), list(x = as.name(n_col))))
  dots2 <- list(.ndocs = substitute(n_distinct(x), list(x = as.name(document_col))))
  dots3 <- list(tf = substitute(x / .document_total, list(x = as.name(n_col))),
                idf = substitute(log(total_documents / .ndocs)))

  ret <- tbl %>%
    group_by_(document_col) %>%
    mutate_(.dots = dots) %>%
    group_by_(term_col) %>%
    mutate_(.dots = dots2) %>%
    ungroup() %>%
    mutate_(.dots = dots3) %>%
    mutate(tf_idf = tf * idf) %>%
    select(-.ndocs, -.document_total)

  if (!is.null(g)) {
    ret <- group_by_(ret, .dots = g)
  }

  ret
}
```

**Comment by juliasilge** [comment]
I'm really glad you've started work on this because tf-idf is one of the things that I have been wanting to explore with a tidy-principle approach. 👍🏻

**Comment by yosuke-yasuda** [comment]
Hi,

I tried calculate_tf_idf using movie_review data from text2vec but it seems to be very slow. The code below takes very very long time.

``` r
library(text2vec)
data("movie_review")
processed_df <- (
  movie_review %>%
    unnest_tokens(output=token, input=review) %>%
    group_by(id, token) %>%
    summarize(count=n()) %>%
    ungroup()
)
calculate_tf_idf(processed_df, token, id, count)
```

I tried using cast_sparse and some text2vec functions and it was much faster.
Why don't you use text2vec? Maybe because its api still unstable?

Even if so, I think you can use it as a reference. The implementation of the calculation is very simple.

I'll put examples from it.

``` r
# text2vec::get_idf 
function (dtm, log_scale = log, smooth_idf = TRUE) 
{
    cs <- colSums(abs(sign(dtm)))
    if (smooth_idf) 
        idf <- log_scale(nrow(dtm)/(cs + 1))
    else idf <- log_scale(nrow(dtm)/(cs))
    Diagonal(dim(dtm)[[2]], idf)
}
```

``` r
# text2vec::transform_tfidf
function (dtm, idf = NULL, sublinear_tf = FALSE, norm = c("l1", 
    "l2")) 
{
    if (!inherits(dtm, "dgCMatrix")) 
        dtm <- as(dtm, "dgCMatrix")
    tf <- transform_tf(dtm, sublinear_tf, norm)
    if (!inherits(idf, "ddiMatrix")) {
        message("idf scaling matrix not provided, calculating it form input matrix")
        idf <- get_idf(dtm)
    }
    tf %*% idf
}
```

**Comment by juliasilge** [comment]
Thanks for trying this out, @double-y! We are still working on the implementation of tf-idf and will look into this.

**Comment by znmeb** [comment]
There are a few packages in CRAN that can do the tf-idf and other statistical NLP tasks - the [Natural Language Processing](https://cran.rstudio.com/web/views/NaturalLanguageProcessing.html) task view lists them.

**Comment by juliasilge** [comment]
I think there is a problem with how term frequency is implemented right now (idf looks good to me). I do NOT think this is related to the new version of dplyr because I saw this before but had not had time to sit down and puzzle it out. The value for tf returned by `calculate_tf_idf` is not the same as what I get just using straight dplyr, for some but not all of the words. I put some examples to show what I mean at the end of the [tf-idf vignette](https://github.com/juliasilge/tidytext/blob/master/vignettes/tf_idf.Rmd) I am working on, if you want to knit it yourself, @dgrtwo. I have looked a bit at the function but I am not seeing what the problem is yet.
